compiler/rustc_codegen_llvm/src/back/archive.rs

```diff
@@ -332,22 +406,46 @@ impl<'a> LlvmArchiveBuilder<'a> {
         }
     }
 
-    fn i686_decorated_name(import: &DllImport) -> CString {
+    fn i686_decorated_name(import: &DllImport, mingw: bool) -> String {
         let name = import.name;
-        // We verified during construction that `name` does not contain any NULL characters, so the
-        // conversion to CString is guaranteed to succeed.
-        CString::new(match import.calling_convention {
-            DllCallingConvention::C => format!("_{}", name),
-            DllCallingConvention::Stdcall(arg_list_size) => format!("_{}@{}", name, arg_list_size),
+        let prefix = if mingw { "" } else { "_" };
+
+        match import.calling_convention {
+            DllCallingConvention::C => format!("{}{}", prefix, name),
+            DllCallingConvention::Stdcall(arg_list_size) => {
+                format!("{}{}@{}", prefix, name, arg_list_size)
+            }
             DllCallingConvention::Fastcall(arg_list_size) => format!("@{}@{}", name, arg_list_size),
             DllCallingConvention::Vectorcall(arg_list_size) => {
                 format!("{}@@{}", name, arg_list_size)
             }
-        })
-        .unwrap()
+        }
     }
 }
 
 fn string_to_io_error(s: String) -> io::Error {
     io::Error::new(io::ErrorKind::Other, format!("bad archive: {}", s))
 }
+
+fn find_binutils_dlltool(sess: &Session) -> OsString {
+    if sess.host.llvm_target != sess.target.llvm_target {
```

Comment:
Porting @nagisa's comment from https://github.com/rust-lang/rust/pull/88801#discussion_r711032367:
> `llvm_target` is probably not the best way to check for this. There are many llvm target triples that are actually equivalent in the end. (e.g. `i686-pc-windows-gnu` and `i686-unknown-windows-gnu` and e.g. json targets could specify either).

How about `if !cfg!(windows) {` like below? Is that too crude?

Reply:
@mati865, @nagisa 

I'll admit I'm a little fuzzy on exactly what this test is trying to achieve, but it looks to me like it's trying to detect a situation in which, for example, we're running rustc on an x86_64-pc-windows-gnu machine but building for i686-pc-windows-gnu, or vice versa.  If that's correct, I don't think `if !cfg!(windows)` would be sufficient, would it?

Of course, this raises the question of what _is_ the right check.  If all we're interested in doing is detecting the case where the host & target architectures differ, should I instead be comparing `sess.host.arch` and `sess.target.arch`?

It may also be relevant to say that I haven't tested these cross-compilation scenarios at all, as I've only succeeded in getting myself confused with all of the various `--build` & `--target` options and what `x.py test` actually seems to run.  To cut down on the number of variables involved, I've fallen back on 4 different checkouts of my branch, one for each of {i686,x86_64}-pc-windows-{msvc,gnu}, each of which sets `build` in Config.toml to the appropriate triple.  (That's intended as context; I'd rather not get distracted by a protracted effort to debug my cross-compilation setup unless it's crucial.)